connectors/sources/sharepoint.py

```diff
@@ -0,0 +1,715 @@
+                "tenant",
+                "tenant_id",
+            ]
+            if self.is_cloud
+            else ["host_url", "site_collections", "username", "password"]
+        )
+
+        if empty_connection_fields := [
+            field for field in connection_fields if self.configuration[field] == ""
+        ]:
+            raise Exception(
+                f"Configured keys: {empty_connection_fields} can't be empty."
+            )
+        if not self.ssl_disabled and self.certificate == "":
+            raise Exception("Configured key ssl_ca can't be empty when SSL is enabled")
+
+    @retryable(
+        retries=RETRIES,
+        interval=RETRY_INTERVAL,
+        strategy=RetryStrategy.EXPONENTIAL_BACKOFF,
+    )
+    async def _set_access_token(self):
+        """Set access token using configuration fields"""
+        if not is_expired(expires_at=self.token_expires_in):
+            return
+        tenant_id = self.configuration["tenant_id"]
+        logger.info("Generating access token...")
+        url = f"https://accounts.accesscontrol.windows.net/{tenant_id}/tokens/OAuth/2"
+        data = {
+            "grant_type": "client_credentials",
+            "resource": f"00000003-0000-0ff1-ce00-000000000000/{self.configuration['tenant']}.sharepoint.com@{tenant_id}",
+            "client_id": f"{self.configuration['client_id']}@{tenant_id}",
+            "client_secret": self.configuration["secret_id"],
+        }
+        headers = {"Content-Type": "application/x-www-form-urlencoded"}
+
+        async with aiohttp.request(
+            method="POST", url=url, data=data, headers=headers
+        ) as response:
+            json_data = await response.json()
+            self.access_token = json_data["access_token"]
+            self.token_expires_in = get_expires_at(int(json_data["expires_in"]))
+
+    async def _generate_session(self):
+        """Generate base client session using configuration fields
+
+        Returns:
+            ClientSession: Base client session.
+        """
+        logger.info("Generating aiohttp Client Session...")
+        request_headers = {
+            "accept": "application/json",
+            "content-type": "application/json",
+        }
+        timeout = aiohttp.ClientTimeout(total=None)
+
+        if self.is_cloud:
+            basic_auth = None
+            await self._set_access_token()
+        else:
+            basic_auth = aiohttp.BasicAuth(
+                login=self.configuration["username"],
+                password=self.configuration["password"],
+            )
+        return aiohttp.ClientSession(
+            auth=basic_auth,
+            headers=request_headers,
+            timeout=timeout,
+            raise_for_status=True,
+        )
+
+    async def close(self):
+        """Closes unclosed client session"""
+        if self.session is None:
+            return
+        await self.session.close()
+
+    async def _api_call(self, url_name, url="", **url_kwargs):
```

Comment:
Can `@retryable` decorator be used here?

Reply:
We need to handle different exceptions here, that's why we can't use the decorator here.